util/billing.go

```diff
@@ -0,0 +1,154 @@
+package util
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/olivere/elastic/v7"
+)
+
+var TierValidity int64
+
+const (
+	envEsURL       = "ES_CLUSTER_URL"
+	arcIdentifier  = "ARC_ID"
+	emailID        = "EMAIL"
+	subscriptionID = "SUBSCRIPTION_ID"
+)
+
+// Middleware function, which will be called for each request
+func BillingMiddleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		currentTime := time.Now().Unix()
+		// Valid plan
+		if TierValidity > currentTime {
+			remainingTime := TierValidity - currentTime
+			// Check if remaining time is less than 24 hrs
+			if remainingTime <= 3600*24 {
+				// Print waring message if remaining time is less than 24 hrs
+				log.Println("warning: payment required. arc will start sending out error messages in next", remainingTime/3600, "hours")
+			}
+			next.ServeHTTP(w, r)
+		} else {
+			// Write an error and stop the handler chain
+			http.Error(w, "payment required", http.StatusPaymentRequired)
+		}
+	})
+}
+
+type ArcUsage struct {
+	ArcID          string `json:"arc_id"`
+	Timestamp      int64  `json:"timestamp"`
+	SubscriptionID string `json:"subscription_id"`
+	Quantity       int    `json:"quantity"`
+	Email          string `json:"email"`
+}
+
+type ArcUsageResponse struct {
+	Accepted      bool   `json:"accepted"`
+	FailureReason string `json:"failure_reason"`
+	ErrorMsg      string `json:"error_msg"`
+	WarningMsg    string `json:"warning_msg"`
+	StatusCode    int    `json:"status_code"`
+	TierValidity  int64  `json:"tier_validity"`
+}
+
+func ReportUsageRequest(arcUsage ArcUsage) (ArcUsageResponse, error) {
+	response := ArcUsageResponse{}
+	url := "https://accapi.appbase.io/arc/" + arcUsage.ArcID + "/report_usage"
+	marshalledRequest, err := json.Marshal(arcUsage)
+	if err != nil {
+		log.Println("error while marshalling req body: ", err)
+		return response, err
+	}
+	req, _ := http.NewRequest("POST", url, bytes.NewBuffer(marshalledRequest))
+	req.Header.Add("Content-Type", "application/json")
+	req.Header.Add("cache-control", "no-cache")
+
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		log.Println("error while sending reequest: ", err)
```

Comment:
s/reequest/request

Reply:
Adding something like [this](https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker) can help spot these.